service_linux.go

```diff
@@ -112,9 +116,149 @@ func (sb *sandbox) populateLoadbalancers(ep *endpoint) {
+			var err error
+			ep, err = n.getEndpointFromStore(e.ID())
+			if err != nil {
+				return nil, nil, err
+			}
+			break
+		}
+	}
+	if ep == nil {
+		return nil, nil, fmt.Errorf("Unable to find load balancing endpoint for network %s", n.ID())
+	}
+	// Get the load balancer sandbox itself as well
+	sb, ok := ep.getSandbox()
+	if !ok {
+		return nil, nil, fmt.Errorf("Unable to get sandbox for %s(%s) in for %s", ep.Name(), ep.ID(), n.ID())
+	}
+	// XXX really?  really?  ugh
+	ep = sb.getEndpoint(ep.ID())
+	if ep == nil {
+		return nil, nil, fmt.Errorf("Load balancing endpoint %s(%s) removed from %s", ep.Name(), ep.ID(), n.ID())
+	}
+	return ep, sb, nil
+}
+
+// Searches the OS sandbox for the name of the endpoint interface
+// within the sandbox.   This is required for adding/removing IP
+// aliases to the interface.
+func findIfaceDstName(sb *sandbox, ep *endpoint) string {
+	srcName := ep.Iface().SrcName()
+	for _, i := range sb.osSbox.Info().Interfaces() {
+		if i.SrcName() == srcName {
+			return i.DstName()
+		}
+	}
+	return ""
+}
+
 // Add loadbalancer backend to all sandboxes which has a connection to
 // this network. If needed add the service as well.
-func (n *network) addLBBackend(ip, vip net.IP, lb *loadBalancer, ingressPorts []*PortConfig) {
+func (n *network) addLBBackend(ip net.IP, lb *loadBalancer) {
+	if len(lb.vip) == 0 {
+		return
+	}
+	switch n.LoadBalacingMode() {
+	case LBModeEndpoint:
+		n.addLBBackendToEndpoint(ip, lb)
+	case LBModeMesh:
+		n.addLBBackendToMesh(ip, lb)
+	default:
+		logrus.Errorf("addLBBackend: network %s is in unknown load balancer mode (%s)!", n.ID(), n.LoadBalacingMode())
+	}
+}
+
+func (n *network) addLBBackendToEndpoint(ip net.IP, lb *loadBalancer) {
+	ep, sb, err := n.findLBEndpointSandbox()
+	if err != nil {
+		logrus.Errorf("error in addLBBackend for %s/%s for %v", n.ID(), n.Name(), err)
+		return
+	}
+	if sb.osSbox == nil {
+		return
+	}
+	if n.ingress && !sb.ingress {
+		return
+	}
+
+	eIP := ep.Iface().Address()
+
+	i, err := ipvs.New(sb.Key())
+	if err != nil {
+		logrus.Errorf("Failed to create an ipvs handle for sbox %s (%s,%s) for lb addition: %v", sb.ID()[0:7], sb.ContainerID()[0:7], sb.Key(), err)
+		return
+	}
+	defer i.Close()
+
+	s := &ipvs.Service{
+		AddressFamily: nl.FAMILY_V4,
+		FWMark:        lb.fwMark,
+		SchedName:     ipvs.RoundRobin,
+	}
+
+	if !i.IsServicePresent(s) {
+		// Add IP alias for the VIP to the endpoint
+		ifName := findIfaceDstName(sb, ep)
+		if ifName == "" {
+			logrus.Errorf("Failed find interface name for endpoint %s(%s) to create LB alias", ep.ID(), ep.Name())
+			return
+		}
+		err := sb.osSbox.AddAliasIP(ifName, &net.IPNet{IP: lb.vip, Mask: net.CIDRMask(32, 32)})
+		if err != nil {
+			logrus.Errorf("Failed add IP alias %s to network %s LB endpoint interface %s: %v", lb.vip, n.ID(), ifName, err)
+			return
+		}
+
+		if sb.ingress {
+			var gwIP net.IP
+			if ep := sb.getGatewayEndpoint(); ep != nil {
+				gwIP = ep.Iface().Address().IP
```

Comment:
Ah, except for the fact that the gwIP address needs to live outside of the scope of the if statement and should be its default value if  `sb.getGatewayEndpoint()` returns `nil`.  It's something akin to writing:
```go
var x int = 0
if something {
    x = 1
}
// Do something with 'x'
```
Except that golint complains if you initialize a newly declared variable to the default value.  In C it would look much nicer of course.  :)
```C
int x = something ? 1 : 0;
```